blockchain/utxobackend.go

```diff
@@ -405,65 +541,86 @@ func (l *LevelDbUtxoBackend) dbPutUtxoBackendInfo(dbTx database.Tx,
 	}
 
 	// Store the database version.
-	meta := dbTx.Metadata()
-	bucket := meta.Bucket(utxoDbInfoBucketName)
-	err := bucket.Put(utxoDbInfoVersionKeyName, uint32Bytes(info.version))
+	err := tx.Put(utxoDbInfoVersionKey, uint32Bytes(info.version))
 	if err != nil {
 		return err
 	}
 
 	// Store the compression version.
-	err = bucket.Put(utxoDbInfoCompressionVerKeyName, uint32Bytes(info.compVer))
+	err = tx.Put(utxoDbInfoCompVerKey, uint32Bytes(info.compVer))
 	if err != nil {
 		return err
 	}
 
 	// Store the UTXO set version.
-	err = bucket.Put(utxoDbInfoUtxoVerKeyName, uint32Bytes(info.utxoVer))
+	err = tx.Put(utxoDbInfoUtxoVerKey, uint32Bytes(info.utxoVer))
 	if err != nil {
 		return err
 	}
 
 	// Store the database creation date.
-	return bucket.Put(utxoDbInfoCreatedKeyName,
+	return tx.Put(utxoDbInfoCreatedKey,
 		uint64Bytes(uint64(info.created.Unix())))
 }
 
-// dbFetchUtxoBackendInfo uses an existing database transaction to fetch the
-// backend versioning and creation information.
-func (l *LevelDbUtxoBackend) dbFetchUtxoBackendInfo(dbTx database.Tx) *UtxoBackendInfo {
-	meta := dbTx.Metadata()
-	bucket := meta.Bucket(utxoDbInfoBucketName)
-
-	// Uninitialized state.
-	if bucket == nil {
-		return nil
-	}
-
+// dbFetchUtxoBackendInfo fetches the backend versioning and creation
+// information.
+func (l *LevelDbUtxoBackend) dbFetchUtxoBackendInfo() (*UtxoBackendInfo, error) {
 	// Load the database version.
-	var version uint32
-	versionBytes := bucket.Get(utxoDbInfoVersionKeyName)
-	if versionBytes != nil {
-		version = byteOrder.Uint32(versionBytes)
+	prefix := utxoPrefixDbInfo
+	versionBytes, err := l.Get(utxoDbInfoVersionKey)
+	if err != nil {
+		return nil, err
+	}
+	if versionBytes == nil {
+		// If the database info was not found, attempt to find it in the legacy
+		// bucket.
+		dbInfoLegacyBucketName := []byte("dbinfo")
+		dbInfoBucketId, err := fetchLegacyBucketId(l.Get, dbInfoLegacyBucketName)
```

Comment:
This code to read the database info (which contains the overall database version) needs to check the old location to detect that it's on an older database version and therefore needs to upgrade.  Since the overall database version always needs to be read in order to determine if upgrades need to be run, I think this code will need to remain as-is.

Reply:
That's fair.  What I was thinking is that the overall database version would act as a way for the upgrade code to know it needs to perform the migration, which it is able to do since it independently understands the buckets and such.  Then, once it's done, this would no longer be needed because there would never be a case where it actually needed to directly load the old information in the normal paths.

However, that would couple the utxodb version to the overall database version which this approach avoids, so doing it this way is indeed better from that standpoint.